File: python/src/webex_message_handler/handler.py
"""Main WebexMessageHandler orchestrator."""

from __future__ import annotations

import asyncio
from collections.abc import Callable
from typing import TYPE_CHECKING, Any

import aiohttp

from .device_manager import DeviceManager
from .kms_client import KmsClient
from .logger import Logger, noop_logger
from .mercury_socket import MercurySocket
from .message_decryptor import MessageDecryptor
from .types import (
    ConnectionStatus,
    DecryptedMessage,
    DeletedMessage,
    DeviceRegistration,
    FetchFunction,
    FetchRequest,
    FetchResponse,
    HandlerStatus,
    InjectedWebSocket,
    MercuryActivity,
    WebexMessageHandlerConfig,
    WebSocketFactory,
)

if TYPE_CHECKING:
    pass

import base64

# Type alias for event callbacks
EventCallback = Callable[..., Any]


def extract_person_uuid(person_id: str) -> str:
    """Extract the raw UUID from a Webex person ID.

    The Webex REST API returns base64-encoded IDs like:
        "Y2lzY29zcGFyazovL3VzL1BFT1BMRS9mYjUx..." → "ciscospark://us/PEOPLE/fb51254f-..."

    Mercury wire format uses raw UUIDs:
        "fb51254f-3b37-4e50-aa04-45744c2effc7"

    This function normalizes both formats to the raw UUID for comparison.
    """
    try:
        decoded = base64.b64decode(person_id).decode("utf-8")
        if decoded.startswith("ciscospark://"):
            uuid = decoded.rsplit("/", 1)[-1]
            if uuid:
                return uuid
    except Exception:
        # Not base64 — treat as raw UUID
        pass
    return person_id


class WebexMessageHandler:
    """Receives and decrypts Webex messages over Mercury WebSocket.

    Usage::

        handler = WebexMessageHandler(WebexMessageHandlerConfig(token="..."))

        @handler.on("message:created")
        async def on_message(msg: DecryptedMessage):
            print(f"[{msg.person_email}] {msg.text}")

        await handler.connect()
    """

    def __init__(self, config: WebexMessageHandlerConfig) -> None:
        if not config.token or not isinstance(config.token, str):
            raise ValueError("WebexMessageHandler requires a non-empty token string")

        # Validate networking mode configuration
        mode = config.mode
        if mode == "injected":
            if not config.fetch or not config.web_socket_factory:
                raise ValueError('Injected mode requires both "fetch" and "web_socket_factory"')
            if config.connector:
                raise ValueError("Cannot use native proxy parameters (connector) in injected mode")
        elif mode == "native":
            if config.fetch or config.web_socket_factory:
                raise ValueError('Cannot provide fetch/web_socket_factory in native mode — set mode to "injected"')
        else:
            raise ValueError(f'Invalid mode "{mode}" — must be "native" or "injected"')

        self._token = config.token
        self._logger: Logger = config.logger or noop_logger  # type: ignore[assignment]
        self._connector = config.connector

        # Create adapters based on mode
        if mode == "native":
            self._http_do = self._create_native_http_adapter(config.connector)
            self._ws_factory = self._create_native_ws_adapter(config.connector)
        else:
            # injected mode - use provided functions
            self._http_do = config.fetch  # type: ignore[assignment]
            self._ws_factory = config.web_socket_factory  # type: ignore[assignment]

        self._device_manager = DeviceManager(
            logger=self._logger,
            http_do=self._http_do,
        )
        self._mercury_socket = MercurySocket(
            logger=self._logger,
            ws_factory=self._ws_factory,
            ping_interval=config.ping_interval,
            pong_timeout=config.pong_timeout,
            reconnect_backoff_max=config.reconnect_backoff_max,
            max_reconnect_attempts=config.max_reconnect_attempts,
        )

        self._ignore_self_messages = config.ignore_self_messages
        self._bot_person_id: str | None = None

        self._kms_client: KmsClient | None = None
        self._message_decryptor: MessageDecryptor | None = None
        self._registration: DeviceRegistration | None = None
        self._connected = False
        self._connecting = False

        # Event listeners
        self._listeners: dict[str, list[EventCallback]] = {
            "message:created": [],
            "message:deleted": [],
            "connected": [],
            "disconnected": [],
            "reconnecting": [],
            "error": [],
        }

        self._setup_mercury_listeners()

    def _create_native_http_adapter(
        self, connector: aiohttp.BaseConnector | None
    ) -> FetchFunction:
        """Create HTTP adapter using native aiohttp."""
        async def http_do(request: FetchRequest) -> FetchResponse:
            async with aiohttp.ClientSession(connector=connector) as session:
                async with session.request(
                    request.method,
                    request.url,
                    headers=request.headers,
                    data=request.body,
                ) as response:
                    # Create a simple response wrapper
                    class NativeFetchResponse:
                        def __init__(self, resp: aiohttp.ClientResponse):
                            self.status = resp.status
                            self.ok = 200 <= resp.status < 300
                            self._response = resp

                        async def json(self) -> Any:
                            return await self._response.json()

                        async def text(self) -> str:
                            return await self._response.text()

                    return NativeFetchResponse(response)  # type: ignore[return-value]

        return http_do

    def _create_native_ws_adapter(
        self, connector: aiohttp.BaseConnector | None
    ) -> WebSocketFactory:
        """Create WebSocket adapter using native aiohttp."""
        async def ws_factory(url: str) -> InjectedWebSocket:
            session = aiohttp.ClientSession(connector=connector)
            ws = await session.ws_connect(url)

            # Attach session for cleanup
            ws._session = session  # type: ignore[attr-defined]

            return ws  # type: ignore[return-value]

        return ws_factory

    def on(self, event: str, callback: EventCallback | None = None) -> Any:
        """Register an event listener. Can be used as a decorator.

        Usage as decorator::

            @handler.on("message:created")
            async def on_message(msg):
                ...

        Usage as method::

            handler.on("message:created", my_callback)
        """
        if callback is not None:
            if event not in self._listeners:
                self._listeners[event] = []
            self._listeners[event].append(callback)
            return callback

        # Decorator usage
        def decorator(func: EventCallback) -> EventCallback:
            if event not in self._listeners:
                self._listeners[event] = []
            self._listeners[event].append(func)
            return func

        return decorator

    def off(self, event: str, callback: EventCallback) -> None:
        """Remove an event listener."""
        listeners = self._listeners.get(event, [])
        if callback in listeners:
            listeners.remove(callback)

    def _emit(self, event: str, *args: Any) -> None:
        """Emit an event to all registered listeners."""
        for callback in self._listeners.get(event, []):
            try:
                result = callback(*args)
                if asyncio.iscoroutine(result):
                    asyncio.ensure_future(result)
            except Exception as exc:
                self._logger.error(f"Error in {event} listener: {exc}")

    async def connect(self) -> None:
        """Establish the full connection pipeline.

        1. Register device with WDM
        2. Connect Mercury WebSocket
        3. Initialize KMS (ECDH handshake)
        4. Begin receiving encrypted messages
        """
        if self._connecting:
            raise RuntimeError("connect() already in progress")
        if self._connected:
            raise RuntimeError("Already connected. Call disconnect() first, or use reconnect(new_token).")

        self._logger.info("Connecting to Webex...")
        self._connecting = True

        try:
            # Step 1: Register device with WDM
            self._registration = await self._device_manager.register(self._token)
            self._logger.info("Device registered")

            # Step 1.5: Fetch bot person info if self-message filtering is enabled
            if self._ignore_self_messages:
                await self._fetch_bot_person_id()

            # Step 2: Create KMS client
            self._kms_client = KmsClient(
                token=self._token,
                device_url=self._registration.device_url,
                user_id=self._registration.user_id,
                encryption_service_url=self._registration.encryption_service_url,
                logger=self._logger,
                http_do=self._http_do,
            )

            # Step 3: Connect Mercury WebSocket FIRST (KMS responses arrive here)
            await self._mercury_socket.connect(
                self._registration.web_socket_url,
                self._token,
            )
            self._logger.info("Mercury connected")

            # Step 4: Initialize KMS (ECDH handshake — response via Mercury)
            await self._kms_client.initialize()
            self._logger.info("KMS initialized")

            # Step 5: Create message decryptor
            self._message_decryptor = MessageDecryptor(
                kms_client=self._kms_client,
                logger=self._logger,
            )

            self._connecting = False
            self._connected = True
            self._logger.info("Connected to Webex")
            self._emit("connected")

        except Exception:
            self._connecting = False
            raise

    async def disconnect(self) -> None:
        """Tear down the connection cleanly."""
        self._logger.info("Disconnecting from Webex...")
        self._connected = False

        await self._mercury_socket.disconnect()

        if self._registration:
            try:
                await self._device_manager.unregister(self._token)
                self._logger.info("Device unregistered")
            except Exception as exc:
                self._logger.warning(f"Failed to unregister device: {exc}")

        self._registration = None
        self._kms_client = None
        self._message_decryptor = None
        self._bot_person_id = None

    async def reconnect(self, new_token: str) -> None:
        """Update the access token and re-establish the connection.

        Tears down the existing connection and reconnects from scratch.
        """
        if not new_token or not isinstance(new_token, str):
            raise ValueError("reconnect() requires a non-empty token string")

        self._logger.info("Reconnecting with new token...")
        await self.disconnect()
        self._token = new_token
        await self.connect()

    @property
    def connected(self) -> bool:
        """Whether the handler is fully connected."""
        return self._connected and self._mercury_socket.connected

    def status(self) -> HandlerStatus:
        """Return a structured health check of all connection subsystems."""
        reconnect_attempt = self._mercury_socket.current_reconnect_attempts

        if self._connected and self._mercury_socket.connected:
            status: ConnectionStatus = "connected"
        elif self._connecting:
            status = "connecting"
        elif reconnect_attempt > 0:
            status = "reconnecting"
        else:
            status = "disconnected"

        return HandlerStatus(
            status=status,
            web_socket_open=self._mercury_socket.connected,
            kms_initialized=self._kms_client is not None,
            device_registered=self._registration is not None,
            reconnect_attempt=reconnect_attempt,
        )

    async def _fetch_bot_person_id(self) -> None:
        """Fetch the bot's person ID for self-message filtering.

        Raises on failure — connect() will not proceed without a valid bot ID
        when ignore_self_messages is enabled.
        """
        self._logger.debug("Fetching bot person info for self-message filtering")
        response = await self._http_do(
            FetchRequest(
                url="https://webexapis.com/v1/people/me",
                method="GET",
                headers={
                    "Authorization": f"Bearer {self._token}",
                    "Content-Type": "application/json",
                },
            )
        )
        if not response.ok:
            raise RuntimeError(
                f"Failed to fetch bot identity for self-message filtering: HTTP {response.status}. "
                "Set ignore_self_messages=False to skip this check (not recommended — may cause message loops)."
            )
        data = await response.json()
        raw_id = data.get("id", "")
        self._bot_person_id = extract_person_uuid(raw_id)
        self._logger.info(f"Bot person ID cached for self-message filtering: {self._bot_person_id}")

    def _setup_mercury_listeners(self) -> None:
        # Forward KMS messages from Mercury to the KMS client
        def on_kms_response(data: dict[str, Any]) -> None:
            if self._kms_client:
                self._kms_client.handle_kms_message(data)

        self._mercury_socket.on("kms:response", on_kms_response)

        # Handle conversation activities
        def on_activity(activity: MercuryActivity) -> None:
            asyncio.ensure_future(self._handle_activity_safe(activity))

        self._mercury_socket.on("activity", on_activity)

        # Handle Mercury reconnection
        def on_mercury_connected() -> None:
            asyncio.ensure_future(self._on_reconnect())

        self._mercury_socket.on("connected", on_mercury_connected)

        # Forward disconnected events
        def on_disconnected(reason: str) -> None:
            self._connected = False
            self._emit("disconnected", reason)

        self._mercury_socket.on("disconnected", on_disconnected)

        # Forward reconnecting events
        def on_reconnecting(attempt: int) -> None:
            self._emit("reconnecting", attempt)

        self._mercury_socket.on("reconnecting", on_reconnecting)

        # Forward error events
        def on_error(err: Exception) -> None:
            self._emit("error", err)

        self._mercury_socket.on("error", on_error)

    async def _handle_activity_safe(self, activity: MercuryActivity) -> None:
        try:
            await self._handle_activity(activity)
        except Exception as exc:
            self._logger.error(f"Error handling activity: {exc}")
            self._emit("error", exc if isinstance(exc, Exception) else Exception(str(exc)))

    async def _handle_activity(self, activity: MercuryActivity) -> None:
        # message:created — verb=post + objectType=comment
        if activity.verb == "post" and activity.object.object_type == "comment":
            if not self._message_decryptor:
                self._logger.warning("Received activity but decryptor not initialized")
                return

            decrypted = await self._message_decryptor.decrypt_activity(activity)
            message = DecryptedMessage(
                id=decrypted.object.id,
                room_id=decrypted.target.id,
                person_id=decrypted.actor.id,
                person_email=decrypted.actor.email_address or "",
                text=decrypted.object.display_name or "",
                html=decrypted.object.content,
                created=decrypted.published,
                room_type=self._infer_room_type(decrypted),
                raw=decrypted,
            )
            # Filter self-messages if enabled
            if self._ignore_self_messages and self._bot_person_id and extract_person_uuid(message.person_id) == self._bot_person_id:
                self._logger.debug(f"Ignoring self-message from bot ({self._bot_person_id})")
                return

            self._emit("message:created", message)
            return

        # message:deleted — verb=delete + objectType=activity
        if activity.verb == "delete" and activity.object.object_type == "activity":
            self._emit(
                "message:deleted",
                DeletedMessage(
                    message_id=activity.object.id,
                    room_id=activity.target.id,
                    person_id=activity.actor.id,
                ),
            )

    @staticmethod
    def _infer_room_type(activity: MercuryActivity) -> str | None:
        tags = activity.target.tags
        if not tags:
            return None
        if "ONE_ON_ONE" in tags:
            return "direct"
        if "TEAM" in tags or "LOCKED" in tags or "GROUP" in tags:
            return "group"
        return None

    async def _on_reconnect(self) -> None:
        self._logger.info("Mercury reconnected, refreshing device and KMS")

        try:
            if self._registration:
                self._registration = await self._device_manager.refresh(self._token)
        except Exception as exc:
            self._logger.warning(f"Device refresh on reconnect failed: {exc}")

        try:
            if self._kms_client:
                await self._kms_client.initialize()
        except Exception as exc:
            self._logger.warning(f"KMS re-init on reconnect failed: {exc}")

        self._connected = True
        self._emit("connected")

File: rust/src/handler.rs
//! Main orchestrator: device registration, Mercury WebSocket, KMS, decryption.

use crate::device_manager::DeviceManager;
use crate::errors::WebexError;
use crate::kms_client::{KmsClient, KmsResponseHandler};
use crate::mercury_socket::{MercuryEvent, MercurySocket};
use crate::message_decryptor::MessageDecryptor;
use crate::types::{
    Config, ConnectionStatus, DecryptedMessage, DeletedMessage, DeviceRegistration, HandlerStatus,
    MercuryActivity,
};
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::{mpsc, Mutex};
use tracing::{error, info, warn};

/// Events emitted by WebexMessageHandler.
#[derive(Debug, Clone)]
pub enum HandlerEvent {
    /// A new message was received and decrypted.
    MessageCreated(DecryptedMessage),
    /// A message was deleted.
    MessageDeleted(DeletedMessage),
    /// Successfully connected (or reconnected).
    Connected,
    /// Disconnected with a reason string.
    Disconnected(String),
    /// Reconnecting (attempt number).
    Reconnecting(u32),
    /// An error occurred.
    Error(String),
}

/// Receives and decrypts Webex messages over Mercury WebSocket.
pub struct WebexMessageHandler {
    token: Arc<Mutex<String>>,
    device_manager: Arc<Mutex<DeviceManager>>,
    mercury_socket: Arc<MercurySocket>,
    kms_client: Arc<Mutex<Option<KmsClient>>>,
    /// Separate handle for resolving KMS responses without locking kms_client.
    kms_response_handler: Arc<Mutex<Option<KmsResponseHandler>>>,
    registration: Arc<Mutex<Option<DeviceRegistration>>>,
    connected: Arc<Mutex<bool>>,
    connecting: Arc<Mutex<bool>>,

    #[allow(dead_code)]
    config: Config,
    event_tx: mpsc::UnboundedSender<HandlerEvent>,
    event_rx: Arc<Mutex<Option<mpsc::UnboundedReceiver<HandlerEvent>>>>,
}

impl WebexMessageHandler {
    /// Create a new WebexMessageHandler.
    pub fn new(config: Config) -> Result<Self, WebexError> {
        if config.token.is_empty() {
            return Err(WebexError::Internal(
                "WebexMessageHandler requires a non-empty token string".into(),
            ));
        }

        let mercury_socket = MercurySocket::new(
            Duration::from_secs_f64(config.ping_interval),
            Duration::from_secs_f64(config.pong_timeout),
            Duration::from_secs_f64(config.reconnect_backoff_max),
            config.max_reconnect_attempts,
        );

        let (event_tx, event_rx) = mpsc::unbounded_channel();

        Ok(Self {
            token: Arc::new(Mutex::new(config.token.clone())),
            device_manager: Arc::new(Mutex::new(DeviceManager::new())),
            mercury_socket: Arc::new(mercury_socket),
            kms_client: Arc::new(Mutex::new(None)),
            kms_response_handler: Arc::new(Mutex::new(None)),
            registration: Arc::new(Mutex::new(None)),
            connected: Arc::new(Mutex::new(false)),
            connecting: Arc::new(Mutex::new(false)),
            config,
            event_tx,
            event_rx: Arc::new(Mutex::new(Some(event_rx))),
        })
    }

    /// Take the event receiver. Can only be called once.
    pub async fn take_event_rx(&self) -> Option<mpsc::UnboundedReceiver<HandlerEvent>> {
        self.event_rx.lock().await.take()
    }

    /// Connect to Webex (register device, connect Mercury, init KMS).
    pub async fn connect(&self) -> Result<(), WebexError> {
        {
            let connecting = self.connecting.lock().await;
            if *connecting {
                return Err(WebexError::Internal("connect() already in progress".into()));
            }
        }
        {
            let connected = self.connected.lock().await;
            if *connected {
                return Err(WebexError::Internal(
                    "Already connected. Call disconnect() first, or use reconnect().".into(),
                ));
            }
        }

        info!("Connecting to Webex...");
        *self.connecting.lock().await = true;

        let result = self.connect_internal().await;

        *self.connecting.lock().await = false;

        match result {
            Ok(()) => {
                *self.connected.lock().await = true;
                info!("Connected to Webex");
                let _ = self.event_tx.send(HandlerEvent::Connected);
                Ok(())
            }
            Err(e) => Err(e),
        }
    }

    async fn connect_internal(&self) -> Result<(), WebexError> {
        let token = self.token.lock().await.clone();

        // Step 1: Register device with WDM
        let reg = {
            let mut dm = self.device_manager.lock().await;
            dm.register(&token).await?
        };
        info!("Device registered");

        // Step 2: Create KMS client
        let kms = KmsClient::new(
            &token,
            &reg.device_url,
            &reg.user_id,
            &reg.encryption_service_url,
        );

        // Get the response handler BEFORE storing the KMS client so the
        // event loop can resolve pending requests without locking kms_client.
        let response_handler = kms.response_handler();
        *self.kms_response_handler.lock().await = Some(response_handler);
        *self.kms_client.lock().await = Some(kms);

        // Step 3: Connect Mercury WebSocket (KMS responses arrive here)
        self.mercury_socket
            .connect(&reg.web_socket_url, &token)
            .await?;
        info!("Mercury connected");

        // Step 4: Start Mercury event loop
        self.start_mercury_event_loop().await;

        // Step 5: Initialize KMS (ECDH handshake — response comes via Mercury)
        {
            let mut kms_guard = self.kms_client.lock().await;
            if let Some(ref mut kms) = *kms_guard {
                kms.initialize().await?;
            }
        }
        info!("KMS initialized");

        // Store registration
        *self.registration.lock().await = Some(reg);

        Ok(())
    }

    /// Start processing Mercury events in a background task.
    async fn start_mercury_event_loop(&self) {
        let mut mercury_rx = match self.mercury_socket.take_event_rx().await {
            Some(rx) => rx,
            None => {
                warn!("Mercury event receiver already taken");
                return;
            }
        };

        let kms_client = self.kms_client.clone();
        let kms_response_handler = self.kms_response_handler.clone();
        let event_tx = self.event_tx.clone();
        let connected = self.connected.clone();
        let registration = self.registration.clone();
        let device_manager = self.device_manager.clone();
        let token = self.token.clone();

        tokio::spawn(async move {
            while let Some(event) = mercury_rx.recv().await {
                match event {
                    MercuryEvent::KmsResponse(data) => {
                        // Use the separate response handler to avoid deadlock
                        // with kms_client lock (held during initialize/get_key).
                        let handler_guard = kms_response_handler.lock().await;
                        if let Some(ref handler) = *handler_guard {
                            handler.handle_kms_message(&data).await;
                        }
                    }
                    MercuryEvent::Activity(activity) => {
                        // Spawn in a separate task so the event loop can continue
                        // processing KMS responses (needed for key retrieval during decryption).
                        let kms_client_clone = kms_client.clone();
                        let event_tx_clone = event_tx.clone();
                        tokio::spawn(async move {
                            let mut kms_guard = kms_client_clone.lock().await;
                            if let Some(ref mut kms) = *kms_guard {
                                Self::handle_activity_static(kms, &activity, &event_tx_clone).await;
                            } else {
                                warn!("Received activity but KMS client not initialized");
                            }
                        });
                    }
                    MercuryEvent::Connected => {
                        info!("Mercury reconnected, refreshing device and KMS");

                        // Refresh device
                        let tok = token.lock().await.clone();
                        {
                            let reg_guard = registration.lock().await;
                            if reg_guard.is_some() {
                                let dm = device_manager.lock().await;
                                match dm.refresh(&tok).await {
                                    Ok(new_reg) => {
                                        drop(reg_guard);
                                        *registration.lock().await = Some(new_reg);
                                    }
                                    Err(e) => {
                                        warn!("Device refresh on reconnect failed: {e}");
                                    }
                                }
                            }
                        }

                        // Re-init KMS
                        {
                            let mut kms_guard = kms_client.lock().await;
                            if let Some(ref mut kms) = *kms_guard {
                                if let Err(e) = kms.initialize().await {
                                    warn!("KMS re-init on reconnect failed: {e}");
                                }
                            }
                        }

                        *connected.lock().await = true;
                        let _ = event_tx.send(HandlerEvent::Connected);
                    }
                    MercuryEvent::Disconnected(reason) => {
                        *connected.lock().await = false;
                        let _ = event_tx.send(HandlerEvent::Disconnected(reason));
                    }
                    MercuryEvent::Reconnecting(attempt) => {
                        let _ = event_tx.send(HandlerEvent::Reconnecting(attempt));
                    }
                    MercuryEvent::Error(msg) => {
                        let _ = event_tx.send(HandlerEvent::Error(msg));
                    }
                }
            }
        });
    }

    /// Handle a single activity (decrypt and route).
    async fn handle_activity_static(
        kms: &mut KmsClient,
        activity: &MercuryActivity,
        event_tx: &mpsc::UnboundedSender<HandlerEvent>,
    ) {
        // message:created — verb=post + objectType=comment
        if activity.verb == "post" && activity.object.object_type == "comment" {
            let mut decryptor = MessageDecryptor::new(kms);
            match decryptor.decrypt_activity(activity).await {
                Ok(decrypted) => {
                    let msg = DecryptedMessage {
                        id: decrypted.object.id.clone(),
                        room_id: decrypted.target.id.clone(),
                        person_id: decrypted.actor.id.clone(),
                        person_email: decrypted
                            .actor
                            .email_address
                            .clone()
                            .unwrap_or_default(),
                        text: decrypted.object.display_name.clone().unwrap_or_default(),
                        html: decrypted.object.content.clone(),
                        created: decrypted.published.clone(),
                        room_type: infer_room_type(&decrypted),
                        raw: decrypted,
                    };
                    let _ = event_tx.send(HandlerEvent::MessageCreated(msg));
                }
                Err(e) => {
                    error!("Error decrypting activity: {e}");
                    let _ = event_tx.send(HandlerEvent::Error(e.to_string()));
                }
            }
            return;
        }

        // message:deleted — verb=delete + objectType=activity
        if activity.verb == "delete" && activity.object.object_type == "activity" {
            let _ = event_tx.send(HandlerEvent::MessageDeleted(DeletedMessage {
                message_id: activity.object.id.clone(),
                room_id: activity.target.id.clone(),
                person_id: activity.actor.id.clone(),
            }));
        }
    }

    /// Disconnect from Webex.
    pub async fn disconnect(&self) {
        info!("Disconnecting from Webex...");
        *self.connected.lock().await = false;

        self.mercury_socket.disconnect().await;

        let token = self.token.lock().await.clone();
        {
            let reg = self.registration.lock().await;
            if reg.is_some() {
                let mut dm = self.device_manager.lock().await;
                if let Err(e) = dm.unregister(&token).await {
                    warn!("Failed to unregister device: {e}");
                } else {
                    info!("Device unregistered");
                }
            }
        }

        *self.registration.lock().await = None;
        *self.kms_client.lock().await = None;
        *self.kms_response_handler.lock().await = None;
    }

    /// Update the access token and re-establish the connection.
    pub async fn reconnect(&self, new_token: &str) -> Result<(), WebexError> {
        if new_token.is_empty() {
            return Err(WebexError::Internal(
                "reconnect() requires a non-empty token string".into(),
            ));
        }

        info!("Reconnecting with new token...");
        self.disconnect().await;

        *self.token.lock().await = new_token.to_string();
        self.connect().await
    }

    /// Whether the handler is fully connected.
    pub async fn connected(&self) -> bool {
        let conn = *self.connected.lock().await;
        conn && self.mercury_socket.connected().await
    }

    /// Returns a structured health check of all connection subsystems.
    pub async fn status(&self) -> HandlerStatus {
        let reconnect_attempt = self.mercury_socket.current_reconnect_attempts().await;
        let ws_open = self.mercury_socket.connected().await;
        let is_connected = *self.connected.lock().await;
        let is_connecting = *self.connecting.lock().await;

        let status = if is_connected && ws_open {
            ConnectionStatus::Connected
        } else if is_connecting {
            ConnectionStatus::Connecting
        } else if reconnect_attempt > 0 {
            ConnectionStatus::Reconnecting
        } else {
            ConnectionStatus::Disconnected
        };

        HandlerStatus {
            status,
            web_socket_open: ws_open,
            kms_initialized: self.kms_client.lock().await.is_some(),
            device_registered: self.registration.lock().await.is_some(),
            reconnect_attempt,
        }
    }
}

fn infer_room_type(activity: &MercuryActivity) -> Option<String> {
    let tags = &activity.target.tags;
    if tags.contains(&"ONE_ON_ONE".to_string()) {
        return Some("direct".to_string());
    }
    if tags.contains(&"TEAM".to_string())
        || tags.contains(&"LOCKED".to_string())
        || tags.contains(&"GROUP".to_string())
    {
        return Some("group".to_string());
    }
    None
}
